File: disciple/src/main/java/com/monster/disciple/CacheBuilder.java
package com.monster.disciple;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CacheBuilder {

    private CacheBuilder() {
    }

    public static CacheBuilder newBuilder() {
        return new CacheBuilder();
    }

    private String cacheKey;

    private final List<Type> args = new ArrayList<>();

    private Type type;

    public CacheBuilder withCacheKey(String cacheKey) {
        this.cacheKey = cacheKey;
        return this;
    }

    public CacheBuilder addType(Class... classes) {
        if (null == classes) {
            throw new NullPointerException("addType() must not null Class");
        }
        args.addAll(Arrays.asList(classes));
        return this;
    }

    public CacheBuilder build() {

        if (1 == args.size()) {
            type = args.get(0);
        } else {
            type = newInstanceType(args, 0);
        }
        return this;
    }

    public Type getType() {
        return type;
    }

    public String getCacheKey() {
        return cacheKey;
    }

    private Type newInstanceType(List<Type> args, int index) {
        if (index + 1 == args.size() - 1) {
            return new ParameterizedTypeImpl((Class) args.get(index), new Type[]{args.get(index + 1)});
        }
        return new ParameterizedTypeImpl((Class) args.get(index), new Type[]{newInstanceType(args, index + 1)});
    }

    private static class ParameterizedTypeImpl implements ParameterizedType {

        private Class raw;
        private Type[] args;

        ParameterizedTypeImpl(Class raw, Type[] args) {
            this.raw = raw;
            this.args = null != args ? args : new Type[0];
        }

        @Override
        public Type[] getActualTypeArguments() {
            return args;
        }

        @Override
        public Type getRawType() {
            return raw;
        }

        @Override
        public Type getOwnerType() {
            return null;
        }
    }
}

File: disciple/src/main/java/com/monster/disciple/util/CacheBuilder.kt
package com.monster.disciple.util

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type
import java.util.ArrayList

class CacheBuilder private constructor() {

    var cacheKey: String? = null
        private set

    private val args = ArrayList<Type>()

    var type: Type? = null
        private set

    fun withCacheKey(cacheKey: String): CacheBuilder {
        this.cacheKey = cacheKey
        return this
    }

    fun addType(vararg classes: Class<*>): CacheBuilder {
        args.addAll(classes)
        return this
    }

    fun build(): CacheBuilder {

        if (1 == args.size) {
            type = args[0]
        } else {
            type = newInstanceType(args, 0)
        }
        return this
    }

    private fun newInstanceType(args: List<Type>, index: Int): Type {
        return if (index + 1 == args.size - 1) {
            ParameterizedTypeImpl(args[index] as Class<*>, arrayOf(args[index + 1]))
        } else ParameterizedTypeImpl(args[index] as Class<*>, arrayOf(newInstanceType(args, index + 1)))
    }

    private class ParameterizedTypeImpl internal constructor(private val raw: Class<*>, args: Array<Type>?) : ParameterizedType {
        private val args: Array<Type> = args ?: emptyArray()

        override fun getActualTypeArguments(): Array<Type> {
            return args
        }

        override fun getRawType(): Type {
            return raw
        }

        override fun getOwnerType(): Type? {
            return null
        }
    }

    companion object {

        fun newBuilder(): CacheBuilder {
            return CacheBuilder()
        }
    }
}
